Add scope support for automatic mixed precision

MattConley opened the pull request:
Allows users to manually prevent certain parts of the graph from being processed by the mixed precision grappler pass

MattConley committed:
Add scope toggle for mixed precision graph rewrite

MattConley committed:
Update Scope API

MattConley committed:
Update to address comments

MattConley commented:
@reedwm

reedwm commented on tensorflow/python/training/experimental/mixed_precision.py:
```diff
@@ -27,6 +31,44 @@
+
+@tf_export(v1=['train.experimental.auto_mixed_precision_scope'])
+@contextlib.contextmanager
+def auto_mixed_precision_scope(select=True):
```
For consistency with `enable_mixed_precision_graph_rewrite`, maybe call this `mixed_precision_graph_rewrite_scope`?

Or, if you don't need to nest these scopes, maybe just have this disable mixed precision, without having the option to set it to True. Then it can be called `disable_mixed_precision_scope()`, or maybe `keep_as_float32`.

@azaks2 any suggestions for the name here? It will go under API review.

MattConley replied:
Sounds good, I changed it to `mixed_precision_graph_rewrite_scope` for now.  If the API without a boolean argument is preferable, I can also add an `enable_mixed_precision_scope()` so we could keep the ability to nest.

reedwm commented on tensorflow/python/training/experimental/mixed_precision.py:
```diff
@@ -27,6 +31,44 @@
 
 
+
+@tf_export(v1=['train.experimental.auto_mixed_precision_scope'])
```
This should be exposed for v2 as well.

MattConley replied:
Thanks, done

reedwm commented on tensorflow/python/ops/gradients_util.py:
```diff
@@ -386,25 +386,48 @@ def _SymGrad(op, out_grads):
 
 
 def _MaybeCompile(scope, op, func, grad_fn):
```
I would rename this to `_run_with_attr_scope`. Or even better, refactor to have this function just return the attribute dictionary, and call it something like `get_extra_grad_attrs`. Then have the caller create the attr_scope and run `grad_fn`.

MattConley replied:
Thanks for the suggestion; I refactored the function to just return the attribute dictionary, let me know if I should make any changes.

MattConley committed:
Address reviewer comments

 -Export scopes for v2
 -Update function name
 -Refactor gradient attribute handling

MattConley commented:
Thanks for the review, just pushed some changes addressing the comments

MattConley committed:
Merge upstream master and address conflicts

MattConley committed:
Update V2 API

MattConley commented:
@reedwm Just removed the V2 exposure and addressed merge conflicts.

reedwm commented:
/CC @alextp, can you give API approval? (Better to give it now than internally, since than @MattConley can respond to feedback).

Also note, I can fix minor lint errors when importing which for some reason aren't being reported.

alextp commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -1062,6 +1063,14 @@ bool AutoMixedPrecisionImpl::IsOnGPU(const NodeDef& node) const {
 
+// Returns whether the node has been disabled via auto_mixed_precision_scope.
+bool AutoMixedPrecisionImpl::IsScopeDisabled(const NodeDef& node) const {
+  if (node.attr().count("_AutoMixedPrecisionScopeInclude")) {
```
I really worry about having these magical attributes; it's very easy for them to go missing when the graph is generated / edited and likewise it's easy for them to get accidentally copied to unintentional places when making graphs for exporting/etc.

Can we use a more proper scoping mechanism, like an op to call a function without converting?

MattConley replied:
You have a good point about the attributes, but I'm not quite sure what you're suggesting as the more proper mechanism.  Could you provide some more detail?

Thanks

rthadur commented:
@alextp any update on this review ?

reedwm commented:
Because the enable_mixed_precision_graph_rewrite API will not be updated for TF 2, and because the tf.keras mixed precision API is recommended for mixed precision training, we do not want scope support to be added for AMP. So I'm closing this PR. I appreciate the work put into this PR, and apologize for not being able to accept it.
